core/src/main/scala/kafka/log/Log.scala

```diff
@@ -1784,8 +1785,24 @@ class Log(@volatile private var _dir: File,
   private def deleteRetentionMsBreachedSegments(): Int = {
     if (config.retentionMs < 0) return 0
     val startMs = time.milliseconds
-    deleteOldSegments((segment, _) => startMs - segment.largestTimestamp > config.retentionMs,
-      reason = s"retention time ${config.retentionMs}ms breach")
+
+    def shouldDelete(segment: LogSegment, nextSegmentOpt: Option[LogSegment]) = {
+      if (startMs - segment.largestTimestamp > config.retentionMs) {
+        segment.largestRecordTimestamp match {
+          case Some(ts) =>
+            info(s"Segment with base offset ${segment.baseOffset} will be deleted due to" +
+              s" retentionMs breach. Largest record timestamp of segment is $ts")
```

Comment:
nit: could we make the connection clearer? How about this?
```scala
   info(s"Segment with base offset ${segment.baseOffset} will be deleted due to" +
              s" retentionMs breach based on the largest record timestamp from the segment, which is $ts")
```

Also, we seem to have lost mention o

Reply:
> Also, we seem to have lost mention o

I'll clarify the log message as you suggested. Could you explain what you meant by the above?

Reply:
Sorry, forgot to finish my thought. I think I was going to mention that we lost the reference to the configuration value.